File: class02/exercise1.py
import numpy
import time
from tqdm import tqdm

PATH = 'data/ex01_python.txt'
N_OF_TRIALS = 30

# n=200までを30回ずつ繰り返し、平均をとったデータをPATHに出力


def matrix_multi(n: int) -> float:
    a = numpy.zeros((n, n))  # Matrix A
    b = numpy.zeros((n, n))  # Matrix B
    c = numpy.zeros((n, n))  # Matrix C

    # Initialize the matrices to some values.
    for i in range(n):
        for j in range(n):
            a[i, j] = i * n + j
            b[i, j] = j * n + i
            c[i, j] = 0

    begin = time.time()

    for i in range(n):
        for j in range(n):
            for k in range(n):
                c[i, j] += a[i][k] * b[k][j]

    end = time.time()
    time_result = end - begin
    # print("time: %.6f sec" % time_result)
    return time_result


if __name__ == '__main__':
    results = []
    # progress bar
    bar = tqdm(total=200)
    # 説明文を追加
    bar.set_description('Progress rate')
    for i in range(2, 201):
        sum = 0.0
        for j in range(N_OF_TRIALS):
            sum += matrix_multi(i)
        # 進捗を設定
        bar.update(1)
        results.append(f'{i}\t{sum/N_OF_TRIALS:e}')

    # esults_str = [format(f, 'e') for f in results]
    with open(PATH, mode='w') as f:
        f.write('\n'.join(results))

File: class02/exercise1.cpp
#include <bits/stdc++.h>
#include <sys/time.h>
using namespace std;
string PATH = "data/ex01_cpp.txt";
int N_OF_TRIALS = 30;

// n=200までを30回ずつ繰り返し、平均をとったデータをPATHに出力

double get_time()
{
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return tv.tv_sec + tv.tv_usec * 1e-6;
}

float matrix_multi(int n)
{
    vector<vector<double>> a(n, vector<double>(n, 0)); // Matrix A
    vector<vector<double>> b(n, vector<double>(n, 0)); // Matrix B
    vector<vector<double>> c(n, vector<double>(n, 0)); // Matrix C

    // Initialize the matrices to some values.
    int i, j, k;
    for (i = 0; i < n; i++)
    {
        for (j = 0; j < n; j++)
        {
            a.at(i).at(j) = i * n + j; // A[i][j]
            b.at(i).at(j) = j * n + i; // B[i][j]
            c.at(i).at(j) = 0;         // C[i][j]
        }
    }

    double begin = get_time();

    for (i = 0; i < n; i++)
    {
        for (j = 0; j < n; j++)
        {
            for (k = 0; k < n; k++)
            {
                c.at(i).at(j) += a.at(i).at(k) * b.at(k).at(j);
            }
        }
    }

    double end = get_time();

    return end - begin;
}

int main()
{
    vector<double> results;
    ofstream outputfile(PATH);

    for (int i = 2; i <= 200; i++)
    {
        double sum = 0;
        for (int j = 0; j < N_OF_TRIALS; j++)
        {
            sum += matrix_multi(i);
        }
        double result = sum / N_OF_TRIALS;
        outputfile << i << "\t" << scientific << result << endl;
    }

    return 0;
}